wee_alloc/src/imp_static_array.rs

```diff
@@ -0,0 +1,86 @@
+use const_init::ConstInit;
+use core::alloc::{AllocErr, Opaque};
+use core::cell::UnsafeCell;
+#[cfg(feature = "extra_assertions")]
+use core::cell::Cell;
+use core::ptr::NonNull;
+use memory_units::{Bytes, Pages};
+use spin::Mutex;
+
+const SCRATCH_LEN_BYTES: usize = 1024 * 1024 * 32;
+static mut SCRATCH_HEAP: [u8; SCRATCH_LEN_BYTES] = [0; SCRATCH_LEN_BYTES];
+static mut OFFSET: Mutex<usize> = Mutex::new(0);
+
+pub(crate) unsafe fn alloc_pages(pages: Pages) -> Result<NonNull<Opaque>, AllocErr> {
+    let bytes: Bytes = pages.into();
+    let mut offset = OFFSET.lock();
+    let end = bytes.0 + *offset;
+    if end < SCRATCH_LEN_BYTES {
+        let ptr = SCRATCH_HEAP[*offset..end].as_mut_ptr() as *mut u8 as *mut Opaque;
+        *offset = end;
+        NonNull::new(ptr).ok_or_else(|| AllocErr)
+    } else {
+        Err(AllocErr)
+    }
+}
+
+pub(crate) struct Exclusive<T> {
+    lock: Mutex<bool>,
+    inner: UnsafeCell<T>,
```

Comment:
This can just be `lock: Mutex<T>`, right? The only reason we aren't using the lock-owns-the-data in other backends is because we don't have the abstractions available (and `Exclusive<T>` is essentially building that abstraction).

Reply:
I believe you're right. The use of a separate `inner` and `lock` is likely unnecessary. IIRC, I kept it that way because `spin::Mutex::new` isn't available by default in `const` context, but now that the `const_fn` feature is being applied, that's not a problem.  Will simplify.